tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11454,6 +11485,38 @@ void DiagnoseEntryAttrAllowedOnStage(clang::Sema *self,
   }
 }
 
+std::vector<FunctionDecl *> GetAllExportedFDecls(clang::Sema *self) {
+  // Add to the end, process from the beginning, to ensure AllExportedFDecls
+  // will contain functions in decl order.
+  std::vector<FunctionDecl *> AllExportedFDecls;
+
+  std::deque<DeclContext *> Worklist;
+  Worklist.push_back(self->getASTContext().getTranslationUnitDecl());
+  while (Worklist.size()) {
+    DeclContext *DC = Worklist.front();
+    Worklist.pop_front();
+    if (auto *FD = dyn_cast<FunctionDecl>(DC)) {
+      if (!FD->hasBody() ||
+          !IsExported(self, FD, getDefaultLinkageExternal(self))) {
```

Comment:
Why is this check necessary?  Isn't it checked before adding a FunctionDecl to the worklist?

Reply:
Originally I thought there would be an edge case where we enter the loop and the first thing in the worklist is a function decl. But, I don't think the translationunitdecl (which starts off the worklist on iteration 0) is ever a function decl, so we are guaranteed it will enter the else case. So, I can see now that the condition isn't necessary. Thanks!